src/server/frontend_xwayland/xwayland_wm_surface.cpp

```diff
@@ -29,6 +29,27 @@ namespace mf = mir::frontend;
 
 namespace
 {
+struct ClientMessageData
+{
+    ClientMessageData(uint32_t const(&data)[5])
+        : data{data}
+    {
+    }
+
+    template<typename T>
+    T unpack()
+    {
+        if (i >= 5)
+            mir::fatal_error("ClientMessageData::unpack() called too many times");
+        auto prev = i;
+        i++;
+        return static_cast<T>(data[prev]);
+    }
+
+    uint32_t const* const data;
+    size_t i{0};
+};
```

Comment:
Seems a needless abstraction that makes the code harder to parse.

Reply:
My requirements are:
- Layout and types for a message defined in one place: all the code that maps the protocol to strongly typed data should be on a few lines, in contrast to spread all over the function.
- No hard-coded indices: these are extremely easy to screw up, and the resulting errors may be difficult to detect
- Doesn't rely on using the correct variant of `i++` vs `++i`: IMO this is almost never good code
- No UB

Do you have a better alternative that satisfies all 4?

Reply:
Well, I don't necessarily agree your requirements.

E.g. I'm not keen on `unpack<>()` having side effects so that the order of calls matters, nor having to look outside the function to see WTF `values.unpack<Action>()` does. A `static_cast<>` is a well known thing with understood risks.

It's just as easy to screw up the the order of unpacking as it is indexes. So I'll just throw in:

```c++
    struct NetWmStateClientMessage
    {
        NetWmStateClientMessage(uint32_t const (&data)[5]) :
            action{static_cast<NetWmStateClientMessageAction>(data[0])},
            properties{static_cast<xcb_atom_t>(               data[1]), 
                       static_cast<xcb_atom_t>(               data[2])},
            source_indication{static_cast<SourceIndication>(  data[3])},
            pad{static_cast<decltype(pad)>(                   data[4])}
        {
            static_assert(sizeof(NetWmStateClientMessage) == sizeof(data), "Structure size incorrect");
        }

        NetWmStateClientMessageAction const action;
        std::array<xcb_atom_t, 2> const properties;
        SourceIndication const source_indication;
        uint32_t const pad;
    } const message{data};
```

Reply:
> It's just as easy to screw up the the order of unpacking as it is indexes

Absolutely not. In the current way you can visually see the order, and so if you know the order properties should be the error is obvious. With indices, you could declare properties in the right order, but unpack the wrong ones or (more likely) copy/paste code but leave duplicate indices. Furthermore, noticing this sort of error is quite hard, even if you're looking at the spec.

Your concern about introducing a non-obvious abstraction are valid. Changed to a more basic method that still satisfies my requirements.